### PR TITLE
Add a G4 material scan script

### PR DESCRIPTION
**Pull request by Chao1009**
### Briefly, what does this PR introduce?
Add a material scan script that utilizes the GEANT4 material scan tool.
It scans over given eta and phi ranges (take the average material thickness over phi).

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No

### Does this PR change default behavior?
No

### CODE REVIEW

**Comment by Chao1009** [comment]
running a test with 81 x 5 = 405 scans.

**Review by wdconinc — CHANGES_REQUESTED** [review]
This doesn't install anywhere. In eic-shell we don't copy the repository but install to a prefix, so this won't be accessible. If we want this user-facing, then it should install. Can you add this to a CMakeLists.txt target?

File: CMakeLists.txt [hunk 1]
```diff
@@ -115,3 +115,8 @@ configure_file(templates/setup.sh.in ${CMAKE_CURRENT_BINARY_DIR}/setup.sh @ONLY)
 install(FILES ${CMAKE_CURRENT_BINARY_DIR}/setup.sh
     DESTINATION ${CMAKE_INSTALL_PREFIX}
 )
+
+# copying test scripts
+install(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/scripts/subdetector_tests
+  DESTINATION ${CMAKE_INSTALL_PREFIX}/scripts
```
Comment (veprbl) [inline CMakeLists.txt:121]:
```suggestion
install(PROGRAMS scripts/subdetector_tests/g4_material_scan.py
  DESTINATION ${CMAKE_INSTALL_BINDIR}
```

File: scripts/subdetector_tests/g4_material_scan.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
+# SPDX-License-Identifier: LGPL-3.0-or-later
```
Comment (veprbl) [inline scripts/subdetector_tests/g4_material_scan.py:1]:
```suggestion
#!/usr/bin/env python3

# SPDX-License-Identifier: LGPL-3.0-or-later
```

File: CMakeLists.txt (hunk 1 above)
Comment (Chao1009) [reply CMakeLists.txt:121]: I think we could put this script in `bin` instead of `scripts/subdetector_tests` instead of installing it as a program

**Comment by wdconinc** [comment]
I would like to avoid confusion between `materialScan`, `g4MaterialScan`, and this script if it is installed in ther user's PATH. It should be clear(er) from the name what it does in addition to these two existing utilities.

**Comment by Chao1009** [comment]
> I would like to avoid confusion between `materialScan`, `g4MaterialScan`, and this script if it is installed in ther user's PATH. It should be clear(er) from the name what it does in addition to these two existing utilities.

Yes, I renamed it to `g4mat_scan_to_csv`.

File: bin/g4mat_scan_to_csv [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        # print(eta_scan.sum(axis=1))
+        # replace nan value with 0 and then calculate average values
+        for mat, xval in eta_scan.fillna(0.).mean(axis=1).items():
+            if mat not in mats_indices:
+                mats_indices[mat] = len(mats_indices)
+            j = mats_indices.get(mat)
+            data[i, j] = xval
+
+    result = pd.DataFrame(columns=mats_indices.keys(), index=etas, data=data[:, :len(mats_indices)])
+    result.to_csv(args.output)
```
Comment (veprbl) [inline bin/g4mat_scan_to_csv:152]:
```suggestion
    result.to_csv(args.output, sep="\t", quoting=csv.QUOTE_NONE)
```
Also, we could switch from CSV to TSV, which is a more common format in HEP

Comment (veprbl) [reply bin/g4mat_scan_to_csv:152]:
```suggestion
    result.to_csv(args.output, sep="\t")
```
`quoting` not supported, it looks like

**Comment by Chao1009** [comment]
The first five materials from the scan
`g4mat_scan_to_csv epic_central.xml --phi 0 --eta 1:2:1 `

          Air  AirOptical  Aluminum  Aluminum5083        Ar
1.0  0.145034    0.000106  0.207398      1.417401  0.000065
2.0  0.329851    0.000000  0.055385      0.000000  0.000000

**Comment by Chao1009** [comment]
Removed averaging in the output, now it has a multi-index: (eta, phi).
One can average the result over phi during the analysis.

**Comment by Chao1009** [comment]
@wdconinc requested changes are addressed.

**Comment by Chao1009** [comment]
> Looks good to me. The only change I'd like you to consider is whether we can rename the script to `g4MaterialScan_to_csv` since it builds on `g4MaterialScan`. That will allow users to find it when they are currently using `g4MaterialScan` by just typing tab. It also makes it clearer that it really is just using `g4MaterialScan` under the hood, so the users will have an accurate understanding of what it does and how it works.
> 
> Similarly, it would be good to use arguments that are similar to `g4MaterialScan`, so a named `--compact` or `-c` flag for the xml file instead of a positional argument. I think this consistency will reduce the possibility of errors.

These are both good suggestions, I've implemented them in the latest commit.

**Review by wdconinc — APPROVED** [review]

**Pull request merged**